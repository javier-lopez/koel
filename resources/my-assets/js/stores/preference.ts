import { userStore } from '.'
import { ls } from '@/services'

export const preferenceStore = {
  storeKey: '',

  state: {
    volume: 7,
    notify: true,
    repeatMode: 'NO_REPEAT',
    showExtraPanel: true,
    confirmClosing: false,
    equalizer: {
      preamp: 0,
      gains: [0, 0, 0, 0, 0, 0, 0, 0, 0, 0]
    },
    artistsViewMode: null,
    albumsViewMode: null,
    selectedPreset: -1,
    transcodeOnMobile: false,
    supportBarNoBugging: false,
    showAlbumArtOverlay: true
  } as { [key: string]: any },

  init (user?: User): void {
    const initUser = user || userStore.current
    this.storeKey = `preferences_${initUser.id}`
    this.state = Object.assign(this.state, ls.get(this.storeKey, this.state))
    this.setupProxy()
  },

  /**
   * Proxy the state properties, so that each can be directly accessed using the key.
   */
  setupProxy (): void {
    Object.keys(this.state).forEach((key: string): void => {
      Object.defineProperty(this, key, {
        get: (): any => this.get(key),
        set: (value: any): void => this.set(key, value),
        configurable: true
      })
    })
  },

  set (key: string, val: any): void {
    this.state[key] = val
    this.save()
  },

  get (key: string): any {
    return key in this.state ? this.state[key] : null
  },

  save (): void {
    ls.set(this.storeKey, this.state)
  }
} as { [key: string]: any }
