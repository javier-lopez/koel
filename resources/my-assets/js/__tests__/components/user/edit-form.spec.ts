import Component from '@/components/user/edit-form.vue'
import { userStore } from '@/stores'
import factory from '@/__tests__/factory'
import { mock } from '@/__tests__/__helpers__'
import { shallow, mount } from '@/__tests__/adapter'

describe('components/user/edit-form', () => {
  afterEach(() => {
    jest.resetModules()
    jest.clearAllMocks()
  })

  it('opens', () => {
    const user = factory('user', {
      name: 'Bob Dylan',
      email: 'knocking@heaven.door'
    })
    const wrapper = shallow(Component, {
      propsData: {
        user
      }
    })
    expect(wrapper).toMatchSnapshot()
  })

  it('saves', () => {
    const user = factory<User>('user', { is_admin: true })
    const updateMock = mock(userStore, 'update')
    const wrapper = shallow(Component, {
      propsData: {
        user
      }
    })
    wrapper.submit('form')
    expect(updateMock).toHaveBeenCalledWith(user, user.name, user.email, user.password, true)
  })

  it('cancels', async () => {
    const user = factory('user')
    const updateMock = mock(userStore, 'update')
    const wrapper = mount(Component, {
      propsData: {
        user
      }
    })

    await wrapper.vm.$nextTick()
    wrapper.click('.btn-cancel')
    expect(wrapper.hasEmitted('close')).toBe(true)
    expect(updateMock).not.toHaveBeenCalled()
  })
})
