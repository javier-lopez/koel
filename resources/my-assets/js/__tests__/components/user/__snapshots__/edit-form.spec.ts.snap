// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/user/edit-form opens 1`] = `
<div class="edit-user">
  <form class="user-edit">
    <header>
      <h1>Edit User</h1>
    </header>
    <div>
      <div class="form-row"><label>Name</label> <input title="Name" type="text" name="name" required="required"></div>
      <div class="form-row"><label>Email</label> <input title="Email" type="email" name="email" required="required"></div>
      <div class="form-row"><label>Password</label> <input name="password" placeholder="Leave blank for no changes" type="password"></div>
      <div class="form-row"><label><input type="checkbox" name="is_admin"> User is an admin
          <i title="Admins can perform administrative tasks like manage users and upload songs." class="fa fa-question-circle help-trigger"></i></label></div>
    </div>
    <footer>
      <btn-stub type="submit" class="btn-update">Update</btn-stub>
      <btn-stub white="" class="btn-cancel">Cancel</btn-stub>
    </footer>
  </form>
</div>
`;
